packages/analytics-core/src/plugins/destination.ts

```diff
@@ -161,53 +164,60 @@ export class Destination implements DestinationPlugin {
         } else {
           this.fulfillRequest(list, res.statusCode, res.status);
         }
-        return;
       }
-      this.handleResponse(res, list);
+      this.handleResponse(res, list, useRetry);
```

Comment:
Distributing the `useRetry` across the multiple handlers makes the logic a little hard to follow.

From what I can tell `useRetry` is needed to determine if diagnostics should be tracked in each `!useRetry` case. However the diagnostics in these cases all seem based on the full list (not retries or drops) and could be done at the top level instead with a separate method. That way we can revert the other response handlers to only handle `useRetry`.

Update lines 166 - 168
```
        processResponseDiagnostics(res, list);
        return;
      }
      this.handleResponse(res, list);
```

New method
```
function processResponseDiagnostics(res, list) {
  const { status } = res;
  switch (status) {
    case Status.Invalid: {
        if (!res.body.error.startsWith(INVALID_API_KEY)) {
          this.config.diagnosticProvider.track(list.length, 400, DIAGNOSTIC_MESSAGES.INVALID_OR_MISSING_FIELDS);
        } else {
            this.config.diagnosticProvider.track(list.length, 400, DIAGNOSTIC_MESSAGES.EVENT_ERROR);
      }
      break;
    }
    case Status.PayloadTooLarge: {
      this.config.diagnosticProvider.track(list.length, 413, DIAGNOSTIC_MESSAGES.PAYLOAD_TOO_LARGE);
      break;
    }
    case Status.RateLimit: { 
      ...
    }
  }
}
```

Reply:
Done in https://github.com/amplitude/Amplitude-TypeScript/pull/568/commits/f52c10d544db3037af15b595763e438dfe835813